In [89]:
import csv
import xlrd
import numpy as np
import pandas as pd

# https://www.census.gov/data/tables/time-series/demo/geographic-mobility/historic.html
annual_geographical_mobility_rates_type_movement = pd.read_excel( '../data/cps-historical-migration-geographic-mobility/tab-a-1.xls' )
annual_inmigration_outmigration_net_migration__movers_from_abroad_for_regions = pd.read_excel( '../data/cps-historical-migration-geographic-mobility/tab-a-2.xls' )
inmigration_outmigration__net_migration_metropolitan_status = pd.read_excel( '../data/cps-historical-migration-geographic-mobility/tab-a-3.xls' )
geographical_mobility_by_tenure = pd.read_excel( '../data/cps-historical-migration-geographic-mobility/tab-a-4.xls' )
reason_for_move_specific_categories = pd.read_excel( '../data/cps-historical-migration-geographic-mobility/tab-a-5.xls' )
distance_of_intercounty_move = pd.read_excel( '../data/cps-historical-migration-geographic-mobility/tab-a-6.xls' )

In [90]:
# annual_geographical_mobility_rates_type_movement
annual_geographical_mobility_rates_type_movement.sample(3)

Table with row headers in column A and column headers in rows 5 through 7.  \
150                                          1965-1966                           
62                                           1969-1970                           
135                                          1985-1986                           

    Unnamed: 1 Unnamed: 2 Unnamed: 3 Unnamed: 4 Unnamed: 5 Unnamed: 6  \
150        100       80.2       19.8       19.3       12.7        6.6   
62      198955     160860      38095      36541      23225      13316   
135        100       81.4       18.6         18       11.3        6.7   

    Unnamed: 7 Unnamed: 8 Unnamed: 9  
150        3.3        3.3        0.5  
62        6250       7066       1554  
135        3.7          3        0.5

annual_geographical_mobility_rates_type_movement

notes:
    seems to cover back to 1947
    needs to be cleaned , pd does not naturally detect formatting 
        top and bottom

In [3]:
# annual_inmigration_outmigration_net_migration__movers_from_abroad_for_regions
annual_inmigration_outmigration_net_migration__movers_from_abroad_for_regions.sample(3)

Table with row headers in column A and column headers in row 5.  \
250                  .Net migration (including abroad)                
28                                        .Outmigrants                
57                             .Net internal migration                

    Unnamed: 1 Unnamed: 2 Unnamed: 3 Unnamed: 4  
250       -120         82       556*       624*  
28         504        575        740        620  
57       -286*         55       240*         -9

annual_inmigration_outmigration_net_migration__movers_from_abroad_for_regions

notes:
    seems to cover back to 1980
    needs to be cleaned , pd does not naturally detect formatting 
        top and bottom

In [4]:
# inmigration_outmigration__net_migration_metropolitan_status
inmigration_outmigration__net_migration_metropolitan_status.sample(3)

Table with row headers in column A and column headers in row 5.  \
185                             .Nonmetropolitan areas                
13                                  ..Principal cities                
156                                                NaN                

    Unnamed: 1 Unnamed: 2 Unnamed: 3 Unnamed: 4 Unnamed: 5  
185       1651       1820       -169        179         10  
13        3196       5515    -2,319*        626    -1,693*  
156        NaN        NaN        NaN        NaN        NaN

inmigration_outmigration__net_migration_metropolitan_status

notes:
    seems to cover back to 1985
    needs to be cleaned , pd does not naturally detect formatting 
        top and bottom

In [5]:
# geographical_mobility_by_tenure
geographical_mobility_by_tenure.sample(3)

Table with row headers in column A and column headers in rows 5 through 7.  Numbers start in row 10, percentages start in row 184, and frequency distributions start in row 358.  \
499                            .Total, 1 year and over                                                                                                                                 
404                         ..In renter-occupied units                                                                                                                                 
271                                                NaN                                                                                                                                 

    Unnamed: 1 Unnamed: 2 Unnamed: 3 Unnamed: 4 Unnamed: 5 Unnamed: 6  \
499        100        100        100        100        100        100   
404       31.3       25.5       71.4       71.4       73.1       67.2   
271        NaN        NaN        NaN        NaN        NaN        NaN   

    Unnamed: 7 Unnamed: 8 Unnamed: 9  
499        100        100        100  
404       67.2       67.2       69.2  
271        NaN        NaN        NaN

geographical_mobility_by_tenure

notes:
    seems to cover back to 1987
    needs to be cleaned , pd does not naturally detect formatting 
        top and bottom

In [6]:
# reason_for_move_specific_categories
reason_for_move_specific_categories.sample(3)

Table with row headers in columns A and K, and column headers in rows 5 and 6. Numbers start in row 9, percentages start in row 36.  \
5                                              NUMBER                                                                                    
23                                                NaN                                                                                    
3                                     Mobility Period                                                                                    

                       Unnamed: 1              Unnamed: 2 Unnamed: 3  \
5                             NaN                     NaN        NaN   
23                            NaN                     NaN        NaN   
3   Total Movers 1 year and over   Family-related reasons        NaN   

   Unnamed: 4                  Unnamed: 5 Unnamed: 6 Unnamed: 7 Unnamed: 8  \
5         NaN                         NaN        NaN        NaN        NaN   
23        NaN                         NaN        NaN        NaN        NaN   
3         NaN  Employment-related reasons        NaN        NaN        NaN   

   Unnamed: 9  ... Unnamed: 12 Unnamed: 13 Unnamed: 14 Unnamed: 15  \
5         NaN  ...         NaN         NaN         NaN         NaN   
23        NaN  ...         NaN         NaN         NaN         NaN   
3         NaN  ...         NaN         NaN         NaN         NaN   

   Unnamed: 16    Unnamed: 17 Unnamed: 18 Unnamed: 19 Unnamed: 20 Unnamed: 21  
5          NaN            NaN         NaN         NaN         NaN         NaN  
23         NaN            NaN         NaN         NaN         NaN         NaN  
3          NaN  Other reasons         NaN         NaN         NaN         NaN  

[3 rows x 22 columns]

reason_for_move_specific_categories

notes:
    seems to cover back to 1998
    needs to be cleaned , pd does not naturally detect formatting 
        top and bottom

In [7]:
# distance_of_intercounty_move
distance_of_intercounty_move.sample(3)

Table with row headers in column A and column headers in row 5. Numbers start in row 8, percentages start in row 32.  \
0   Table A-6. Distance of Intercounty Move: 2003-...                                                                     
24                                        2004-2005/4                                                                     
37                                          2011-2012                                                                     

   Unnamed: 1 Unnamed: 2 Unnamed: 3 Unnamed: 4 Unnamed: 5  
0         NaN        NaN        NaN        NaN        NaN  
24      15287       4657       3752       3007       3870  
37        100    40.2128    20.6384    13.7055    25.4433

distance_of_intercounty_move

notes:
    - seems to cover back to 2002
    - needs to be cleaned , pd does not naturally detect formatting 
        top and bottom

## Aggregate Dates
    - first recording year varies by up to 55 years 
| first recording year | dataset |
|------|------|
| 1947 | annual_geographical_mobility_rates_type_movement |
|------|------|
| 1980 | annual_inmigration_outmigration_net_migration__movers_from_abroad_for_regions |
|------|------|
| 1985 | inmigration_outmigration__net_migration_metropolitan_status |
|------|------|
| 1987 | inmigration_outmigration__net_migration_metropolitan_status |
|------|------|
| 1998 | reason_for_move_specific_categories |
|------|------|
| 2002 | distance_of_intercounty_move |

- let's drop the rows clustered with NaNs

In [8]:
# copy dataframe , generalize name
df = annual_geographical_mobility_rates_type_movement.copy()
# drop columns without any non-NaN values
df = df.dropna(thresh= ( 1 ) )
# return the dataframe to see results
# df
df.sample(7)

Table with row headers in column A and column headers in rows 5 through 7.  \
23                                           2006-2007                           
122                                          1995-1996                           
60                                         1970-1971/8                           
108                                          2005-2006                           
36                                           1997-1998                           
75                                           1958-1959                           
65                                           1966-1967                           

    Unnamed: 1 Unnamed: 2 Unnamed: 3 Unnamed: 4 Unnamed: 5 Unnamed: 6  \
23      292749     254068      38681      37490      25192      12298   
122        100       83.7       16.3       15.8       10.3        5.6   
60      201506     163800      37705      36161      23018      13143   
108        100       86.3       13.7       13.3        8.6        4.7   
36      265209     222702      42507      41304      27082      14222   
75      170658     137018      33640      32804      22315      10489   
65      192233     155710      36523      35200      22339      12861   

    Unnamed: 7 Unnamed: 8 Unnamed: 9  
23        7436       4862       1191  
122        3.1        2.5        0.5  
60        6197       6946       1544  
108        2.8          2        0.4  
36        7867       6355       1203  
75        5419       5070        836  
65        6308       6553       1323

notes:
    - 156 rows remain
    - some column names in row 3 , some column names in row 4 , some column names in row 5
        row 3 = 6 column names
            columns = 0 , 1 , 2 , 3 , 4 , 9
        row 4 = 2 column names
            columns = 5 , 6
        row 5 = 2 column names
            columns = 7 , 8
    - row 4 column 4 and row 5 column 6 contain 'Total'
        not obvious what this is
        but it is not the column name in either case
        in both cases , the column name is one row higher in the same column
    - rows 173 - 185 contain only 1 NaN value

actions:
    - drop all rows without 2 or more non-NaN values
    - merge rows 3 , 4 , and 5 
        taking out 'Total'

In [9]:
# rename dataframe
df = annual_geographical_mobility_rates_type_movement.copy()

# drop columns with less than 2 non-NaN values
df = df.dropna( thresh= ( 2 ) )

# collect column names as pandas series
col_names_0to4_9 = df.iloc[ 0 ]
col_names_5to6 = df.iloc[ 1 ]
col_names7to8 = df.iloc[ 2 ]

# initialize list for column names
column_names = []

# fill list with first row of column names , no exceptions
for column_name in col_names_0to4_9:    
    column_names.append( column_name )

# set index count to -1 so column 0 == column 0
_ = -1

# replace items 5 and 6 in column_names with column names 5 and 6
for column_name in col_names_5to6:
    _ += 1
    # column_name is string (nan instances = float) and name is not 'Total'   
    if isinstance( column_name , str ) and column_name != 'Total':
        # column_name is aligned to replace a nan         
        if str( column_names[ _ ] ) == 'nan':
            # replace item in column_names with column_name             
            column_names[ _ ] = column_name

# set index count to -1 so column 0 == column 0
_ = -1

# replace items 7 and 8 in column_names with column names 7 and 8
for column_name in col_names7to8:
    _ += 1
    # column_name is string (nan instances = float) and name is not 'Total'   
    if isinstance( column_name , str ) and column_name != 'Total':
        # column_name is aligned to replace a nan         
        if str( column_names[ _ ] ) == 'nan':
            # replace item in column_names with column_name             
            column_names[ _ ] = column_name

# reset column names
df.columns = column_names

# drop rows used to extract column names
df = df[3:]

# reset dataframe index
df = df.reset_index( drop=True ) 

# let's see
df.sample(7)

Mobility period Total, 1 year old and over  \
45                     1969-1970                     198955   
130                    1952-1953                        100   
43                   1975-1976/8                     208069   
109                    1981-1982                        100   
35                     1987-1988                     237431   
63                     1951-1952                     150494   
8    2010-2011 (2000 controls)/3                     302005   

    Same residence (non-movers) Total movers  \
45                       160860        38095   
130                        79.4         20.6   
43                       171276        36793   
109                          83           17   
35                       195258        42174   
63                       120016        30478   
8                        266930        35075   

    Different residence in the United States Same county Different county  \
45                                     36541       23225            13316   
130                                     20.1        13.5              6.6   
43                                     35645       22399            13246   
109                                     16.6        10.3              6.2   
35                                     40974       26201            14772   
63                                     29840       19874             9966   
8                                      34016       23325            10691   

    Same state Different state Movers from abroad  
45        6250            7066               1554  
130          3             3.6                0.5  
43        7106            6140               1148  
109        3.3               3                0.5  
35        7727            7046               1200  
63        4854            5112                638  
8         5912            4779               1058

as of now, we have:
    1) removed rows with NaN values
    2) extracted, sorted, and set column names
    
next:
    - we need to clean up the 'Mobility period' column
        - samples from current:
            2017-2018
            2013-2014 (98,000 address file)/1
            2011-2012
            2010-2011 (2010 controls)/2
            2000-2001 (SCHIP, 2000 controls)/5
            1970-1971/8
        - what we want:
            2018 << 2017-2018
            2014 << 2013-2014 (98,000 address file)/1
                requires further investigation
            2012 << 2011-2012
            2011 << 2010-2011 (2010 controls)/2
                requires further investigation
            2001 << 2000-2001 (SCHIP, 2000 controls)/5
                requires further investigation
            1971 << 1970-1971/8

In [10]:
# determine type of data in 'Mobility period' 
def test_type( dataframe , test_type ):
    '''
    input) dataframe w/ specified column 
        - e.g. df['specified column']
    input) type wanting to test
        - e.g. str 
    
    output) count of instances which are not type == test_type
    '''
    non_test_type = 0
    for _ in dataframe:
        if isinstance( _ , test_type ) == False:
            non_test_type += 1
    return non_test_type


test_000 = test_type( dataframe=df[ 'Mobility period' ] , test_type=str )
test_000

0

as test_000 returns 0:
    - the type of all datapoints in 'Mobility period' is str (string)
    - we will not need to change any points to have uniformity 
    
next up:
    - clear low-hanging fruit
    - change the simple ranges to int for set year
        - e.g. change "2017-2018" to "2018"

In [11]:
# identify changes which can be implimented without further investigation
mobility_period = df[ 'Mobility period' ]
simple_year_ranges_count_ = 0
confirmed_simple_year_ranges_count_ = 0

# count how many simple-fix years we have
for datapoint in mobility_period:
    # len('2012-2013') == 9 , check datapoint is formatted as expected     
    if len( datapoint ) == 9 and len( datapoint.split( '-' ) ) == 2:
        # pretty sure we know what's going on but will investigate further         
        simple_year_ranges_count_ += 1
        # first 4 characters of string , can be converted to int
        first_year = int( datapoint[ :4 ] )
        # last 4 characters of string , can be converted to int
        second_year = int( datapoint[ -4: ] )
        # first_year and last_year are post-1940 (because expected), and year range is 1 year 
        if first_year and second_year > 1940 and second_year - first_year == 1:
            # the datapoint is within our risk threshold
            confirmed_simple_year_ranges_count_ += 1

# if everything we thought was what we think it is 
if simple_year_ranges_count_ == confirmed_simple_year_ranges_count_:
    print(f'total: {confirmed_simple_year_ranges_count_}\nratio: {confirmed_simple_year_ranges_count_/len(mobility_period)}')

total: 110
ratio: 0.8088235294117647


notes:
    - know count of lowest fruit to be 110 ( > 80.88% , not bad)
    - just confirmed status of 110 points to be (at least) similar enough to extract years
    - knowing simple_year_ranges_count_ allows the conformation of outcome via count compairson
    
actions:
    - convert those 110 year ranges to years represented by the 2nd year in each range

In [12]:
less_low_mobility_periods = []
# so row 0 == row 0
_ = -1
# little bit of ctrl + c , little bit of ctrl + v
for datapoint in mobility_period:
    _ += 1
    # len('2012-2013') == 9 , check datapoint is formatted as expected     
    if len( datapoint ) == 9 and len( datapoint.split( '-' ) ) == 2:
        # replace datapoint w/ int of last 4 characters of string 
        mobility_period[ _ ] = int( datapoint[ -4: ] )
    else:
        # keep what's already there
        less_low_mobility_periods.append( datapoint )

mobility_period.sample(7)

55                            1960
19     2000-2001 (2000 controls)/3
113                           1970
126                           1957
83                     2004-2005/4
111                    1975-1976/8
17                            2003
Name: Mobility period, dtype: object

- notes
    - we have converted the low hanging fruit

- actions
    - now on to exploring the other 26 rows

In [13]:
# print them out , let's see what we can see
_ = 0
for row in df['Mobility period']:
    _ += 1    
    if not isinstance( row , int ):
        if _ < 10:
            print( str(_) + '  |  ' + row )
        elif _ < 100:
            print( str(_) + ' |  ' + row )
        else:
            print( str(_) + '|  ' + row )

5  |  2013-2014 (98,000 address file)/1
8  |  2010-2011 (2010 controls)/2
9  |  2010-2011 (2000 controls)/3
10 |  2009-2010 (2010 controls)/2
11 |  2009-2010 (2000 controls)/3
16 |  2004-2005/4
20 |  2000-2001 (2000 controls)/3
21 |  2000-2001 (SCHIP, 2000 controls)/5
22 |  2000-2001 (1990 controls)/6
30 |  1992-1993 (1990 controls)/6
31 |  1992-1993 (1980 controls)/7
44 |  1975-1976/8
45 |  1970-1971/8
73 |  2013-2014 (98,000 address file)/1
76 |  2010-2011 (2010 controls)/2
77 |  2010-2011 (2000 controls)/3
78 |  2009-2010 (2010 controls)/2
79 |  2009-2010 (2000 controls)/3
84 |  2004-2005/4
88 |  2000-2001 (2000 controls)/3
89 |  2000-2001 (SCHIP, 2000 controls)/5
90 |  2000-2001 (1990 controls)/6
98 |  1992-1993 (1990 controls)/6
99 |  1992-1993 (1980 controls)/7
112|  1975-1976/8
113|  1970-1971/8


- notes
    - 6 rows formatted almost like majority (new lowest)
        - all have either /4 or /8 at the end of otherwise standard string
            - e.g. standard 
                - 2004-2005
            - observation rows 
                - 16 |  2004-2005/4
                - 44 |  1975-1976/8
                - 45 |  1970-1971/8
                - 84 |  2004-2005/4
                - 112|  1975-1976/8
                - 113|  1970-1971/8
        - remarks before reading documentation
            - identical instances
                - rows 16 and 84
                - rows 44 and 112
                - rows 45 and 113
            - rows are not ordered by year
                - unless years reoccour/loop through table
                - or otherwise
    - 16 rows are in the next new lowest format
        - all contain ({year} controls)/{num} with year and num in consistent pairs
            - (year / num) pairs = # occurrences
                - (2010 / 2) = 4
                - (2000 / 3) = 6
                - (1990 / 6) = 4
                - (1980 / 7) = 2
        - they occour in pattern twice:
            - 4 uninterrupted
            - 1
            - 3 uninterrupted
        - sample rows
            - 8  |  2010-2011 (2010 controls)/2
            - 9  |  2010-2011 (2000 controls)/3
            - 10 |  2009-2010 (2010 controls)/2
            - 11 |  2009-2010 (2000 controls)/3
        - unknown analysis
            - seems to be game settings for different censuses 
            - difficult to connect the /2,/3 and /6,/7 relationship
                - if was dependent on decade order , possibles would be straight ordered 
                    - e.g. /2, /3, /4, /5 or /1, /3, /5, /7
                    
    - final 4 rows are longer and are 2 duplicate pairs
        - default format + ending either (98,000 address file)/1 or (SCHIP, 2000 controls)/5
            - longest rows
                - 5  |  2013-2014 (98,000 address file)/1
                - 21 |  2000-2001 (SCHIP, 2000 controls)/5
                - 73 |  2013-2014 (98,000 address file)/1
                - 89 |  2000-2001 (SCHIP, 2000 controls)/5
        - pre document remarks
            - first notice of /1 
            - return use of /5 
                - do other /n have additional possibile (paramaters)
    - blind analysis
        - year range
                - min: 1970-1971 (1971 in complete df)
                - max: 2013-2014 (2014 in complete df)
            - less range than full representation of column
                - min: 1948 (1947-1948 in origional df)
                - max: 2018 (2017-2018 in origional df)
            - why?
                - recent changes to census
                    - collection
                    - questions
                    - coverage
                - recent increase in census related surveys
                    - housing surveys
                    - business surveys
                        - we will be exploring these after fitting gen_000 model
                        - these are also how this data is presented for non-decade years
        - lot of possible meanings for non-conforming objects
                - politics..
                - changing standards
                - technological evolution
                - etc...
                    - storage , collection , methods , etc... 
            - full-court lobs
                - something to do with immigrants
                - evolution of process 
                - expansion of non-physical opportunities 
        - there's got to be a better way to extract dataframes from excel files
            - likely convertable, but annotations likely direct either universe to same result 
- actions
    - go find out what these non-standard occurrences mean

In [30]:
# make another copy of origional dataframe (this is why we copy)
info_df = annual_geographical_mobility_rates_type_movement.copy()

# function to make NaN dataframe
row_nans = lambda df: df[ df.isnull().any( axis=1 ) ]

# identify the rows with information (basically, keep the NaNs this time)
nan_df = row_nans( info_df )

# peak some rows with NaN values 
nan_df.head()

Table with row headers in column A and column headers in rows 5 through 7.  \
0  Table A-1. Annual Geographical Mobility Rates,...                           
1                            (Numbers in thousands.)                           
2                                                NaN                           
3                                    Mobility period                           
4                                                NaN                           

                   Unnamed: 1                   Unnamed: 2    Unnamed: 3  \
0                         NaN                          NaN           NaN   
1                         NaN                          NaN           NaN   
2                         NaN                          NaN           NaN   
3  Total, 1 year old and over  Same residence (non-movers)  Total movers   
4                         NaN                          NaN           NaN   

                                 Unnamed: 4   Unnamed: 5        Unnamed: 6  \
0                                       NaN          NaN               NaN   
1                                       NaN          NaN               NaN   
2                                       NaN          NaN               NaN   
3  Different residence in the United States          NaN               NaN   
4                                     Total  Same county  Different county   

  Unnamed: 7 Unnamed: 8          Unnamed: 9  
0        NaN        NaN                 NaN  
1        NaN        NaN                 NaN  
2        NaN        NaN                 NaN  
3        NaN        NaN  Movers from abroad  
4        NaN        NaN                 NaN

In [60]:
# noted column 0 name 
column_names = []

# make it slightly educational
for _ in nan_df[ :0 ]:
    # each _ is a string
    if isinstance( _ , str ):
        # add the string to column_names
        column_names.append( _ )
        
# this is the education
print( f'len(nan_df[:0])=={ len( nan_df[ :0 ] ) }\ntype(nan_df[:0])=={ type( nan_df[ :0 ] ) }\n'
      # The economy, stupid. 
       f'len(column_names)=={ len( column_names ) }\ntype(column_names={ type( column_names ) }' )

len(nan_df[:0])==0
type(nan_df[:0])==<class 'pandas.core.frame.DataFrame'>
len(column_names)==10
type(column_names=<class 'list'>


In [99]:
'''back to business'''
# we can find row headers in column A
if column_names[0] != 'Table with row headers in column A and column headers in rows 5 through 7.':
    raise AssertionError('')

# function to return non-NaN rows 
non_nans = lambda df: df.dropna()

# identify non-NaN rows from column A (the first column)
col_a_nn = non_nans( nan_df[ 'Table with row headers in column A and column headers in rows 5 through 7.' ] )

# let's see
col_a_nn

0      Table A-1. Annual Geographical Mobility Rates,...
1                                (Numbers in thousands.)
3                                        Mobility period
6                                                 NUMBER
90                                               PERCENT
173                                           Footnotes:
174    1/ The 2014 CPS ASEC included redesigned quest...
175    2/ Population controls consistent with 2010 Ce...
176    3/ Population controls consistent with 2000 Ce...
177    4/ Caution should be used when comparing numbe...
178    5/ SCHIP sample: poulation controls consistent...
179    6/ Population controls consistent with 1990 Ce...
180    7/ Population controls consistent with 1980 Ce...
181    8/ The 1-year geographic mobility question was...
182    For more information on the transition from Ce...
183    Adjustments to Household Survey Population Est...
184       Population Estimates in January 2012 webpage\n
185    Source:  U.S. Census Bur

In [169]:
# identify information of interest
indicators_of_interest = [ '1/' , '2/' , '3/' , '4/' , '5/' , '6/' , '7/' , '8/' ]

# make copy
copy_col_a_nn = col_a_nn.copy()

# extract information of interest
# for data in copy_col_a_nn:
#     for indicators in data:

for indicator in indicators_of_interest:
    # [x for b in a for x in b]
    data = [ string for row in copy_col_a_nn for string in row if indicator in string]
    print(data)

# name_later(copy_col_a_nn)
# set aside information we are not interested in at this time
# thanks_but_later = [ information for information in copy_col_a_nn if copy_col_a_nn[information] not in rows_of_interest ]

[]
[]
[]
[]
[]
[]
[]
[]
